### PR TITLE
[CMake] Rework build of ASTGen and ObservationMacros

### PR DESCRIPTION
**Pull request by DougGregor**
Refactor the build support for ASTGen and ObservationMacros. Both have nearly-identical copies of a bunch of messy, workaround-laden CMake code to build "pure" Swift host libraries using CMake's Swift support.

Instead, introduce `add_pure_swift_host_library` to centralize all of the hacks to build a host library that's all Swift, using CMake's support directly (rather than custom commands), and link against the swift-syntax stack. Switch ASTGen directly over to this.

Add `add_swift_macro_library` on top of this, to make it easy to create a macro library and install it into the appopriate plugin directory. Switch ObservationMacros over to this.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Review by phausler — APPROVED** [review]
Looks great to me!

**Comment by DougGregor** [comment]
@swift-ci please smoke test

File: stdlib/public/Observation/Sources/Observation/CMakeLists.txt [hunk 1]
```diff
@@ -10,42 +10,25 @@
...
-# which does not build in the same way just yet
-if (${SWIFT_HOST_VARIANT_SDK} IN_LIST SWIFT_DARWIN_PLATFORMS)
-  list(APPEND swift_runtime_library_compile_flags -I${SWIFT_SOURCE_DIR}/stdlib/include -I${SWIFT_SOURCE_DIR}/include)
-
-  set(SWIFT_OBSERVATION_SWIFT_FLAGS)
-  set(SWIFT_OBSERVATION_DEPENDS)
-
-  list(APPEND SWIFT_OBSERVATION_SWIFT_FLAGS
-    "-enable-experimental-feature" "Macros"
-    "-DSWIFT_OBSERVATION_MACROS"
```
Comment (bnbarham) [inline stdlib/public/Observation/Sources/Observation/CMakeLists.txt:23]: We're no longer adding `-DSWIFT_OBSERVATION_MACROS` but the library depends on it, ie.
```
#if SWIFT_OBSERVATION_MACROS
...
#end
```
in `stdlib/public/Observation/Sources/Observation/Macros.swift`. I don't think that check is actually needed, can we just remove it?

**Review by bnbarham — APPROVED** [review]
Thanks, much better :)

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Comment by DougGregor** [comment]
@swift-ci please smoke test

**Pull request merged**